src/components/graph/GraphSecondary.java

```diff
@@ -0,0 +1,393 @@
+package components.graph;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Stack;
+
+/**
+ * Layered on top of {@code GraphKernel} with secondary methods.
+ *
+ * @author Zheng Ni
+ */
+public abstract class GraphSecondary implements Graph {
+
+    /**
+     * The primary representation variable for the graph. It is the core of a
+     * graph representation, including the tags of each vertex and a set
+     * including each vertex's neighbor vertices' tags.
+     */
+    private ArrayList<Vertex> vertices;
```

Comment:
Your abstract class should not have any fields. You should interact with the vertices indirectly using the kernel methods.